test/celeritas/optical/ValidationUtils.hh

```diff
@@ -0,0 +1,186 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/ValidationUtils.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <array>
+#include <type_traits>
+#include <vector>
+
+#include "corecel/cont/Span.hh"
+#include "corecel/data/Collection.hh"
+#include "celeritas/UnitTypes.hh"
+#include "celeritas/grid/GenericGridData.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+
+#include "Test.hh"
+#include "TestMacros.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+namespace test
+{
+using namespace ::celeritas::test;
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of a hard-coded test data array.
+ *
+ * Converts the arguments supplied in units \c UnitType to native units.
+ */
+template<class UnitType, class... Args>
+std::array<real_type, sizeof...(Args)> constexpr expressed_as(Args const&... args)
+{
+    return std::array<real_type, sizeof...(Args)>{
+        native_value_from(UnitType(args))...};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of hard-coded test data.
+ *
+ * Same as \c expressed_as, except returns a vector.
+ */
+template<class UnitType, class... Args>
+std::vector<real_type> vector_expressed_as(Args const&... args)
+{
+    return std::vector<real_type>{native_value_from(UnitType(args))...};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Checks two \c ImportPhysicVector for exact equality.
+ */
+void check_physics_vector(ImportPhysicsVector const& expected,
+                          ImportPhysicsVector const& actual);
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a floating point type collection into a \c real_type span.
+ *
+ * Used to convert data which may be expressed as double or single precision
+ * into the same size as \c real_type. If the supplied data is the same
+ * precision, then just a span to the original data is returned. Otherwise,
+ * it is copied into a local buffer and casted to the correct precision. The
+ * \c RealSpanGenerator object should live as long as its spans are being
+ * used in the tests.
+ */
+class RealSpanGenerator
```

Comment:
Is this class used anywhere? I think we may be spending too much effort supporting single precision. I'm actually fine if the physics tests pass only with double; the single precision mainly helps elucidate edge cases in low-level math functions (rounding error) and geometry. Eventually we may want to switch to separate "setup time" and "runtime" precision levels so that most interfaces are doubles.

Reply:
After peeking at some of the test implementations, I think I managed to refactor this out entirely now.